distributed/scheduler.py

```diff
@@ -7488,36 +7532,92 @@ def decide_worker(
     of bytes sent between workers.  This is determined by calling the
     *objective* function.
     """
-    ws: WorkerState = None
     wws: WorkerState
-    dts: TaskState
+
+    group: TaskGroup = ts._group
+    ws: WorkerState = group._last_worker
+
+    if valid_workers is not None:
+        total_nthreads = sum(wws._nthreads for wws in valid_workers)
+
+    group_tasks_per_worker = len(group) / total_nthreads
```

Comment:
What happens here if `valid_workers is None`?  It looks like total_nthreads might be undefined

Reply:
Then we use the `self._total_nthreads` that `SchedulerState.decide_worker` passed in (see https://github.com/dask/distributed/pull/4899/files#diff-bbcf2e505bf2f9dd0dc25de4582115ee4ed4a6e80997affc7b22122912cc6591R2376). So in the common case of no worker restrictions, we won't have to do the `total_nthreads` calculation.